### PR TITLE
Split fetching device keys and signatures into two transactions

### PR DESCRIPTION
**Pull request by richvdh**
I think this is simpler (and moves stuff out of the db threads)

~~based on #8231~~

### CODE REVIEW

File: synapse/storage/databases/main/end_to_end_keys.py [hunk 1]
```diff
@@ -154,22 +156,57 @@ async def get_e2e_device_keys_and_signatures(
...
-    def _get_e2e_device_keys_and_signatures_txn(
+    def _get_e2e_device_keys_txn(
         self, txn, query_list, include_all_devices=False, include_deleted_devices=False
     ) -> Dict[str, Dict[str, Optional[DeviceKeyLookupResult]]]:
+        """Get information on devices from the database
+
+         The results include the device's keys and self-signatures, but *not* any
+         cross-signing signatures which have been added subsequently (for which, see
+         get_e2e_device_keys_and_signatures)
+         """
```
Comment (clokep) [inline synapse/storage/databases/main/end_to_end_keys.py:207]: Looks like the spacing got weird.

```suggestion
        """Get information on devices from the database

        The results include the device's keys and self-signatures, but *not* any
        cross-signing signatures which have been added subsequently (for which, see
        get_e2e_device_keys_and_signatures)
        """
```

File: synapse/storage/databases/main/end_to_end_keys.py [hunk 2]
```diff
@@ -221,41 +250,32 @@ def _get_e2e_device_keys_and_signatures_txn(
...
+        self, txn: Cursor, device_query: Iterable[Tuple[str, str]]
+    ) -> List[Dict]:
+        """Get cross-signing signatures for a given list of devices
 
-            target_device_signatures = target_device_result.signatures
-            if target_device_signatures is None:
-                target_device_signatures = target_device_result.signatures = {}
+        Returns signatures made by the owner of the devices. Each entry in the result
+        is a dict containing the fields from the database ('user_id', 'key_id',
+        'target_user_id', 'target_device_id', 'signature').
```
Comment (clokep) [inline synapse/storage/databases/main/end_to_end_keys.py:262]: Not sure if it really matters but don't we know that `user_id == target_user_id` here? Seems silly to return them separately (as we're just pumping more bytes over the database connection). But maybe this is trying to micro-optimize.

**Review by clokep — APPROVED** [review]

**Comment by clokep** [comment]
Also, CI seems a bit finicky, but I don't think it is from this PR.

File: synapse/storage/databases/main/end_to_end_keys.py (hunk 2 above)
Comment (richvdh) [reply synapse/storage/databases/main/end_to_end_keys.py]: well you asked for it; it now avoids pulling the spurious column :)

File: synapse/storage/databases/main/end_to_end_keys.py [hunk 3]
```diff
@@ -178,20 +181,14 @@ async def get_e2e_device_keys_and_signatures(
             )
 
             # add each cross-signing signature to the correct device in the result dict.
-            for row in cross_sigs_result:
-                signing_user_id = row["user_id"]
-                signing_key_id = row["key_id"]
-                target_user_id = row["target_user_id"]
-                target_device_id = row["target_device_id"]
-                signature = row["signature"]
```
Comment (clokep) [inline synapse/storage/databases/main/end_to_end_keys.py]: This is much easier to read. 👍

**Review by clokep — APPROVED** [review]

**Pull request merged**